// superagentai-js/dist/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SuperAgentClient = void 0;
const Client_1 = require("./api/resources/agent/client/Client");
const Client_2 = require("./api/resources/llm/client/Client");
const Client_3 = require("./api/resources/apiUser/client/Client");
const Client_4 = require("./api/resources/datasource/client/Client");
const Client_5 = require("./api/resources/tool/client/Client");
const Client_6 = require("./api/resources/workflow/client/Client");
class SuperAgentClient {
    constructor(_options) {
        this._options = _options;
    }
    get agent() {
        var _a;
        return ((_a = this._agent) !== null && _a !== void 0 ? _a : (this._agent = new Client_1.Agent(this._options)));
    }
    get llm() {
        var _a;
        return ((_a = this._llm) !== null && _a !== void 0 ? _a : (this._llm = new Client_2.Llm(this._options)));
    }
    get apiUser() {
        var _a;
        return ((_a = this._apiUser) !== null && _a !== void 0 ? _a : (this._apiUser = new Client_3.ApiUser(this._options)));
    }
    get datasource() {
        var _a;
        return ((_a = this._datasource) !== null && _a !== void 0 ? _a : (this._datasource = new Client_4.Datasource(this._options)));
    }
    get tool() {
        var _a;
        return ((_a = this._tool) !== null && _a !== void 0 ? _a : (this._tool = new Client_5.Tool(this._options)));
    }
    get workflow() {
        var _a;
        return ((_a = this._workflow) !== null && _a !== void 0 ? _a : (this._workflow = new Client_6.Workflow(this._options)));
    }
}
exports.SuperAgentClient = SuperAgentClient;
